gtk4/src/subclass/text_buffer.rs

```diff
@@ -0,0 +1,543 @@
+                f(text_buffer.unsafe_cast_ref::<TextBuffer>().to_glib_none().0)
+            }
+        }
+    }
+}
+
+unsafe impl<T: TextBufferImpl> IsSubclassable<T> for TextBuffer {
+    fn override_vfuncs(class: &mut glib::Class<Self>) {
+        <Object as IsSubclassable<T>>::override_vfuncs(class);
+
+        let klass = class.as_mut();
+        klass.apply_tag = Some(text_buffer_apply_tag::<T>);
+        klass.begin_user_action = Some(text_buffer_begin_user_action::<T>);
+        klass.changed = Some(text_buffer_changed::<T>);
+        klass.delete_range = Some(text_buffer_delete_range::<T>);
+        klass.end_user_action = Some(text_buffer_end_user_action::<T>);
+        klass.insert_child_anchor = Some(text_buffer_insert_child_anchor::<T>);
+        klass.insert_paintable = Some(text_buffer_insert_paintable::<T>);
+        klass.insert_text = Some(text_buffer_insert_text::<T>);
+        klass.mark_deleted = Some(text_buffer_mark_deleted::<T>);
+        klass.mark_set = Some(text_buffer_mark_set::<T>);
+        klass.modified_changed = Some(text_buffer_modified_changed::<T>);
+        klass.paste_done = Some(text_buffer_paste_done::<T>);
+        klass.remove_tag = Some(text_buffer_remove_tag::<T>);
+        klass.redo = Some(text_buffer_redo::<T>);
+        klass.undo = Some(text_buffer_undo::<T>);
+    }
+}
+
+unsafe extern "C" fn text_buffer_apply_tag<T: TextBufferImpl>(
+    ptr: *mut ffi::GtkTextBuffer,
+    tag_ptr: *mut ffi::GtkTextTag,
+    start_ptr: *const ffi::GtkTextIter,
+    end_ptr: *const ffi::GtkTextIter,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<TextBuffer> = from_glib_borrow(ptr);
+
+    imp.apply_tag(
+        &wrap.unsafe_cast_ref(),
+        &from_glib_borrow(tag_ptr),
+        &from_glib_borrow(start_ptr),
+        &from_glib_borrow(end_ptr),
+    )
+}
+
+unsafe extern "C" fn text_buffer_begin_user_action<T: TextBufferImpl>(
+    ptr: *mut ffi::GtkTextBuffer,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<TextBuffer> = from_glib_borrow(ptr);
+
+    imp.begin_user_action(&wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn text_buffer_changed<T: TextBufferImpl>(ptr: *mut ffi::GtkTextBuffer) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<TextBuffer> = from_glib_borrow(ptr);
+
+    imp.changed(&wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn text_buffer_delete_range<T: TextBufferImpl>(
+    ptr: *mut ffi::GtkTextBuffer,
+    start_ptr: *mut ffi::GtkTextIter,
+    end_ptr: *mut ffi::GtkTextIter,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<TextBuffer> = from_glib_borrow(ptr);
+
+    imp.delete_range(
+        &wrap.unsafe_cast_ref(),
+        &mut from_glib_none(start_ptr),
+        &mut from_glib_none(end_ptr),
```

Comment:
This is not sufficient. You need to write back to the pointers afterwards if the iters have changed (also for other mutable text iters in the other functions).

In gtk3 I fixed that somewhere at some point, it's all a bit awful to get right. `gtk_text_iter_assign()` can be useful here.

Reply:
The problem basically is that `from_glib_none()` crates a copy, `from_glib_borrow()` is immutable and not a copy.

Reply:
Aha right, thanks, should be properly handled now